flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/tasks/mailbox/MailboxProcessor.java

```diff
@@ -0,0 +1,319 @@
+		// action becomes available again.
+		while (isDefaultActionUnavailable() && isMailboxLoopRunning()) {
+			mailbox.takeMail().run();
+		}
+
+		return isMailboxLoopRunning();
+	}
+
+	/**
+	 * Calling this method signals that the mailbox-thread should (temporarily) stop invoking the default action,
+	 * e.g. because there is currently no input available.
+	 */
+	private MailboxDefaultAction.SuspendedDefaultAction suspendDefaultAction() {
+
+		assert taskMailboxExecutor.isMailboxThread();
+
+		if (suspendedDefaultAction == null) {
+			suspendedDefaultAction = new SuspendDefaultActionRunnable();
+			ensureControlFlowSignalCheck();
+		}
+
+		return suspendedDefaultAction;
+	}
+
+	private boolean isDefaultActionUnavailable() {
+		return suspendedDefaultAction != null;
+	}
+
+	private boolean isMailboxLoopRunning() {
+		return mailboxLoopRunning;
+	}
+
+	/**
+	 * Helper method to make sure that the mailbox loop will check the control flow flags in the next iteration.
+	 */
+	private void ensureControlFlowSignalCheck() {
+		// Make sure that mailbox#hasMail is true via a dummy letter so that the flag change is noticed.
+		if (!mailbox.hasMail()) {
+			try {
+				mailbox.tryPutMail(() -> {});
+			} catch (MailboxStateException me) {
+				LOG.debug("Mailbox closed when trying to submit letter for control flow signal.", me);
+			}
+		}
+	}
+
+	/**
+	 * Implementation of {@link MailboxDefaultAction.ActionContext} that is connected to a {@link MailboxProcessor}
+	 * instance.
+	 */
+	private final class MailboxDefaultActionContext implements MailboxDefaultAction.ActionContext {
+
+		private final MailboxProcessor mailboxProcessor;
+
+		private MailboxDefaultActionContext(MailboxProcessor mailboxProcessor) {
+			this.mailboxProcessor = mailboxProcessor;
+		}
+
+		@Override
+		public void allActionsCompleted() {
+			mailboxProcessor.allActionsCompleted();
+		}
+
+		@Override
+		public MailboxDefaultAction.SuspendedDefaultAction suspendDefaultAction() {
+			return mailboxProcessor.suspendDefaultAction();
+		}
+	}
+
+	/**
+	 * Represents the suspended state of the default action and offers an idempotent method to resume execution.
+	 */
+	private final class SuspendDefaultActionRunnable implements MailboxDefaultAction.SuspendedDefaultAction {
+
+		/** Ensuring idempotent behavior, we ensure this is only accessed from the main thread. */
+		private boolean valid;
+
+		SuspendDefaultActionRunnable() {
+			this.valid = true;
+		}
+
+		@Override
+		public void resume() {
+			try {
+				if (taskMailboxExecutor.isMailboxThread()) {
+					resumeInternal();
+				} else {
+					mailbox.putMail(this::resumeInternal);
+				}
+			} catch (InterruptedException ie) {
+				Thread.currentThread().interrupt();
+			} catch (MailboxStateException me) {
+				LOG.debug("Action context could not submit letter to mailbox.", me);
+			}
+		}
+
+		private void resumeInternal() {
+			if (valid) {
+				valid = false;
+				suspendedDefaultAction = null;
+			}
```

Comment:
```java
if (suspendedDefaultAction == this) {
    suspendedDefaultAction = null;
}
```
?

Reply:
I think it is better like this, or else client code could accidentally keep a reference on one outdated suspended action forever and it would always be able to manipulate the flag.

Reply:
The suggested snipped should behave exactly the same as it's now in the PR.

>  client code could accidentally keep a reference on one outdated suspended action forever and it would always be able to manipulate the flag.

If by "outdated" you mean a `SuspendDefaultActionRunnable` instance that was already called `resumeInternal()`, then it would set `valid = false` and next invocations of this method have no effect.

On the other hand, you have invariants in the code, that:
 * if `SuspendDefaultActionRunnable.valid`, then it's the same instance that referenced in `MailboxProcessor.suspendedDefaultAction`.
 * if there are multiple instances of `SuspendDefaultActionRunnable` returned by `MailboxProcessor.suspendDefaultAction()`, only the latest one can have `valid = true`.

Reply:
True, so I will remove the flag.